### PR TITLE
Add Theia user preferences synchronizer

### PR DESCRIPTION
**Pull request by mmorhun**
Theia user preferences synchronizer is designed to handle Theia user settings synchronization between workspaces. It retrieves Theia user preferences from Che user preferences on Theia server start and watches changes, so when any update is done it will be reflected back to Che storage.

Depends on [Preferences API PR](https://github.com/eclipse/che-theia/pull/79)

Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### CODE REVIEW

File: extensions/che-theia-user-preferences/src/node/che-theia-preferences-contribition.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from "inversify";
+import { BackendApplicationContribution } from '@theia/core/lib/node/backend-application';
+import { CheTheiaUserPreferencesSynchronizer } from './che-theia-preferences-synchronizer';
+
+@injectable()
+export class CheTheiaPreferencesContrinution implements BackendApplicationContribution {
```
Comment (benoitf) [inline extensions/che-theia-user-preferences/src/node/che-theia-preferences-contribition.ts]: hello, FYI name of the file and name of the class have a typo

name of file should contain `contribUtion` and class `contriBution` (n/b)

Comment (mmorhun) [reply extensions/che-theia-user-preferences/src/node/che-theia-preferences-contribition.ts]: Yes, thanks you! Will fix

**Comment by mmorhun** [comment]
I used extension because we need to bring user preferences on Theia server start.
Another reason is optimization. We retrieve them only once, on Che Theia start instead of doing it on each plug-in start.

File: extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+import { injectable, inject } from "inversify";
+import { readFileSync, writeFileSync } from 'fs';
+import { homedir } from 'os';
+import { sep as PATH_SEPARATOR } from 'path';
+import * as nsfw from 'vscode-nsfw';
+import { CheApiService, Preferences } from "@eclipse-che/theia-plugin-ext/lib/common/che-protocol";
+
+export const THEIA_PREFERENCES_KEY = 'theia-user-preferences';
+export const THEIA_USER_PREFERENCES_PATH = homedir() + PATH_SEPARATOR + '.theia' + PATH_SEPARATOR + 'settings.json';
```
Comment (benoitf) [inline extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts]: FYI : path.resolve()
https://nodejs.org/api/path.html#path_path_resolve_paths
```
path.resolve('wwwroot', 'static_files/png/', '../gif/image.gif');
// If the current working directory is /home/myself/node,
// this returns '/home/myself/node/wwwroot/static_files/gif/image.gif'
```

File: extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+/// <reference types="@theia/core/src/typings/nsfw/index"/>
+
+import { injectable, inject } from "inversify";
```
Comment (benoitf) [inline extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts]: somehow we have a missing formatter/linter as some imports are with double quotes and some other with single quotes

File: extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-contribution.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import { CheTheiaUserPreferencesSynchronizer } from './che-theia-preferences-synchronizer';
+
+@injectable()
+export class CheTheiaPreferencesContribution implements BackendApplicationContribution {
+
+    @inject(CheTheiaUserPreferencesSynchronizer)
+    cheTheiaUserPreferencesSynchronizer: CheTheiaUserPreferencesSynchronizer
+
+    async onStart(): Promise<void> {
+        await this.cheTheiaUserPreferencesSynchronizer.readTheiaUserPreferencesFromCheSettings();
```
Comment (benoitf) [inline extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-contribution.ts]: is it an issue to block the start of theia to wait for the preferences ?

File: extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts (hunk 2 above)
Comment (mmorhun) [reply extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts]: Yes, I had knew it, but do not know why didn't use... will fix. Thanks.

File: extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-contribution.ts (hunk 4 above)
Comment (mmorhun) [reply extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-contribution.ts]: The idea was to provide preferences before Theia start.
I do not mind to remove await here, watcher should do the job.

File: extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts (hunk 3 above)
Comment (mmorhun) [reply extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts]: Indeed. Will check.

Comment (mmorhun) [reply extensions/eclipse-che-theia-user-preferences/src/node/che-theia-preferences-synchronizer.ts]: Done, thanks

File: extensions/eclipse-che-theia-user-preferences/package.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    "che",
+    "preferences"
+  ],
+  "version": "0.0.1",
+  "files": [
+    "lib",
+    "src"
+  ],
+  "dependencies": {
+    "@theia/core": "^0.3.19",
```
Comment (benoitf) [inline extensions/eclipse-che-theia-user-preferences/package.json]: drop of carrets ?

**Pull request merged**